language/move-vm/runtime/src/interpreter.rs

```diff
@@ -869,6 +1033,504 @@ impl Frame {
+            Bytecode::MutBorrowLoc(idx) => {
+                let ty = local_tys[*idx as usize].clone();
+                interpreter
+                    .operand_stack
+                    .push_ty(Type::MutableReference(Box::new(ty)))?;
+            }
+            Bytecode::ImmBorrowLoc(idx) => {
+                let ty = local_tys[*idx as usize].clone();
+                interpreter
+                    .operand_stack
+                    .push_ty(Type::Reference(Box::new(ty)))?;
+            }
+            Bytecode::ImmBorrowField(fh_idx) => {
+                let expected_ty = resolver.field_handle_to_struct(*fh_idx);
+                let top_ty = interpreter.operand_stack.pop_ty()?;
+                top_ty.check_ref_eq(&expected_ty)?;
+                interpreter
+                    .operand_stack
+                    .push_ty(Type::Reference(Box::new(resolver.get_field_type(*fh_idx)?)))?;
+            }
+            Bytecode::MutBorrowField(fh_idx) => {
+                let expected_ty = resolver.field_handle_to_struct(*fh_idx);
+                let top_ty = interpreter.operand_stack.pop_ty()?;
+                top_ty.check_eq(&Type::MutableReference(Box::new(expected_ty)))?;
+                interpreter
+                    .operand_stack
+                    .push_ty(Type::MutableReference(Box::new(
+                        resolver.get_field_type(*fh_idx)?,
+                    )))?;
+            }
+            Bytecode::ImmBorrowFieldGeneric(idx) => {
+                let expected_ty = resolver.field_instantiation_to_struct(*idx, ty_args)?;
+                let top_ty = interpreter.operand_stack.pop_ty()?;
+                top_ty.check_ref_eq(&expected_ty)?;
+                interpreter.operand_stack.push_ty(Type::Reference(Box::new(
+                    resolver.instantiate_generic_field(*idx, ty_args)?,
+                )))?;
+            }
+            Bytecode::MutBorrowFieldGeneric(idx) => {
+                let expected_ty = resolver.field_instantiation_to_struct(*idx, ty_args)?;
+                let top_ty = interpreter.operand_stack.pop_ty()?;
+                top_ty.check_eq(&Type::MutableReference(Box::new(expected_ty)))?;
+                interpreter
+                    .operand_stack
+                    .push_ty(Type::MutableReference(Box::new(
+                        resolver.instantiate_generic_field(*idx, ty_args)?,
+                    )))?;
+            }
+            Bytecode::Pack(idx) => {
+                let field_count = resolver.field_count(*idx);
+                let args_ty = resolver.get_struct_fields(*idx)?;
+                let output_ty = resolver.get_struct_type(*idx);
+                let ability = resolver.loader().abilities(&output_ty)?;
+                let field_abilities = if ability.has_key() {
+                    ability
+                        .remove(Ability::Key)
+                        .union(AbilitySet::singleton(Ability::Store))
+                } else {
+                    ability
```

Comment:
It took me a minute to convince myself that this is correct in the case that the struct being packed has `store`, but not `key`? e.g.,
```
struct A {}
struct B has store {
  a: A
}

let x = B { a: A{} } // should get flagged by runtime checks
```

A comment on why this is safe might be helpful.